test/test_ufbcup.F90

```diff
@@ -0,0 +1,97 @@
+! This is a test for NCEPLIBS-bufr library.
+!
+! This tests the ufbcub() subroutine.
+!
+! Ed Hartnett 4/20/23
+program test_ufbcup
+  implicit none
+  character*8 subset
+  integer jdate, iret
+  real*8 hdr(1, 1)
+  real*8 EPSILON
+  parameter(EPSILON = .01)
+  real*8 missing
+  character*8 station
+  equivalence(station, hdr(1, 1))
+  
+  print *, 'Testing ufbcup.'
+
+#ifdef KIND_8
+  call setim8b(.true.)
+#endif
+
+  ! Open input file.
+  open(unit = 11, file = 'testfiles/IN_9', form ='unformatted')
+  call openbf(11, 'IN', 11)
+
+  ! Open output file.
+  open(unit = 12, file = 'test_ufbcup_out', form ='unformatted')
+  call openbf(12, 'OUT', 11)
+
+  ! Read a message.
+  call readmg(11, subset, jdate, iret)
+  if (iret .ne. 0 .or. subset .ne. 'ADPSFC' .or. jdate .ne. 23022519) stop 10
+
+  ! Load a subset of data.
+  call readsb(11, iret)
+  if (iret .ne. 0) stop 20
+
+  ! Create a new message in the output file.
+  call openmg(12, subset, jdate)
+
+  ! Call ufbcup.
+  call ufbcup(11, 12)
+
+  ! Write the data subset.
+  call writsb(12)
+
+  ! Close files.
+  call closbf(11)
+  call closbf(12)
+
+  ! Now reopen the file as input.
+  open(unit = 12, file = 'test_ufbcup_out', form ='unformatted')
+
+  ! Open the file, and read the 1st subset of the 1st message.
+  call rdmgsb(12, 1, 1)
+
+  ! Get IREC, the number of BUFR messages, according to ufbint()
+  ! documentation. But I get a very large number of messages!
+  call ufbint(12, hdr, 1, 1, iret, 'IREC')
+  if (hdr(1,1) .ne. 100000000000.00000_8) stop 30
+  
+  ! Get SID, the station ID.
+  call ufbint(12, hdr, 1, 1, iret, 'SID')
+  if (station .ne. 'CWGN    ') stop 31
+
+  ! Get the MISSING value.
+  call ufbint(12, hdr, 1, 1, iret, 'NUL')
+  missing = hdr(1, 1)
+  if (abs(hdr(1, 1) - 100000000000.00000_8) > EPSILON) stop 32
+  
+  ! Get and check more values.
```

Comment:
FYI, there's already a library function which you can use to test this sort of thing, rather than having to hardcode to an EPSILON offset from the current library "missing" value.  See the function ibfms, where you just pass in the value you want to compare to "missing", and then it does the rest and returns 0 if the value isn't "missing" and 1 if it is.  We use it all over the place, both inside other library routines, as well as in outside test codes, application programs, etc.

Reply:
BTW, you can also use getbmiss to get the current "missing" value, rather than having to read it yourself out of hdr(1,1)

Reply:
Though in looking through this code, if you use ibfms everywhere then I don't think you need to have a `real*8 missing` variable in the first place, unless I missed something.

Reply:
Fair enough, and I'll go ahead and merge this.  And I'll just open up a new issue to remind us to go back and update this at some point.